src/lib/types.ts

```diff
@@ -209,14 +210,319 @@ type ResolveInviteCodeResponse = {
  * which is further defined below.
  */
-type ApplicationEvent = CalendarCreatedEvent;
 
-type CalendarCreatedEvent = {
-  type: "calendar_created";
+type ApplicationEvent =
+  | UserNameAssigned
+  | CalendarCreated
+  | CalendarUpdated
+  | CalendarDeleted
+  | CalendarAccessRequested
+  | CalendarAccessAccepted
+  | CalendarAccessRejected
+  | CalendarAccessRejected
+  | SpaceCreated
+  | SpaceUpdated
+  | SpaceDeleted
+  | ResourceCreated
+  | ResourceUpdated
+  | ResourceDeleted
+  | EventCreated
+  | EventUpdated
+  | EventDeleted
+  | SpaceRequested
+  | SpaceRequestAccepted
+  | SpaceRequestRejected
+  | SpaceRequestAcceptanceRevoked
+  | ResourceRequested
+  | ResourceRequestAccepted
+  | ResourceRequestRejected
+  | ResourceRequestAcceptanceRevoked
+  | UserRoleAssigned;
+
+/**
+ * Reused types
+ */
+
+type SpaceRequestId = Hash;
+type ResourceRequestId = Hash;
+
+type CalendarFields = {
+  calendarName: string;
+  calendarDates: TimeSpan[];
+};
+
+type SpaceFields = {
+  type: "physical" | "gps" | "virtual";
+  name: string;
+  location: PhysicalLocation | GPSLocation | VirtualLocation;
+  capacity: number;
+  accessibility: string;
+  description: string;
+  contact: string;
+  message: string;
+  link: Link;
+  images: Image[];
+  availability: TimeSpan[] | "always";
+  multiBookable: boolean;
+};
+
+type ResourceFields = {
+  name: string;
+  ownerId: PublicKey;
+  description: string;
+  contact: string;
+  link: Link;
+  images: Image[];
+  availability: TimeSpan[] | "always";
+  multiBookable: boolean; // resource can be booked more than once in the same time span
+};
+
+type EventFields = {
+  name: string;
+  description: string;
+  location?: SpaceRequestId; // ref to a space
+  startDate: Date; // allocated time of a space
+  endDate: Date; // allocated time of a space
+  publicStartDate?: Date; // public facing
+  publicEndDate?: Date; // public facing
+  resources: ResourceRequestId[];
+  links: Link[];
+  images: Image[];
+};
+
+/**
+ * User
+ *
+ * NOTE: sent on `inbox` channel
+ */
+
+type UserNameAssigned = {
+  type: "user_name_assigned";
   data: {
     name: string;
-    startDate?: string;
-    endDate?: string;
+  };
+};
+
+/**
+ * Access
+ *
+ * NOTE: sent on `inbox` channel
+ */
+
+type CalendarAccessRequested = {
+  type: "calendar_access_requested";
+  data: {
+    calendarId: Hash;
+    name: string;
+    message: string;
+  };
+};
+
+type CalendarAccessAccepted = {
+  type: "calendar_access_accepted";
+  data: {
+    calendarId: Hash;
+  };
+};
+
+type CalendarAccessRejected = {
+  type: "calendar_access_rejected";
+  data: {
+    calendarId: Hash;
+  };
+};
+
+/**
+ * Calendar
+ */
+
+type CalendarCreated = {
+  type: "calendar_created";
+  data: {
+    fields: CalendarFields;
+  };
+};
+
+type CalendarUpdated = {
+  type: "calendar_updated";
+  data: {
+    id: Hash;
+    fields: CalendarFields;
+  };
+};
+
+type SpacesPageDescriptionUpdated = {
```

Comment:
These messages update the descriptions of specific pages in the calendar and so we gave them their own message types (some more collision avoidance).

Reply:
I think they all need a `calendarId` to refer to, right?

Reply:
It seems to me we are tying the protocol too close to different ways of having pages in clients. If we add a page we need to change the protocol etc. pp.

How about a generic `PageUpdated` thing with a page type field?